In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px

In [70]:
data = pd.read_excel('get_around_delay_analysis.xlsx')

In [71]:
data

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
0         505000  363965       mobile  canceled                           NaN   
1         507750  269550       mobile     ended                         -81.0   
2         508131  359049      connect     ended                          70.0   
3         508865  299063      connect  canceled                           NaN   
4         511440  313932       mobile     ended                           NaN   
...          ...     ...          ...       ...                           ...   
21305     573446  380069       mobile     ended                           NaN   
21306     573790  341965       mobile     ended                        -337.0   
21307     573791  364890       mobile     ended                         144.0   
21308     574852  362531      connect     ended                         -76.0   
21309     575056  351549      connect     ended                          35.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
0                           NaN                                         NaN  
1                           NaN                                         NaN  
2                           NaN                                         NaN  
3                           NaN                                         NaN  
4                           NaN                                         NaN  
...                         ...                                         ...  
21305                  573429.0                                       300.0  
21306                       NaN                                         NaN  
21307                       NaN                                         NaN  
21308                       NaN                                         NaN  
21309                       NaN                                         NaN  

[21310 rows x 7 columns]

In [72]:
canceled = data[data['state'] == 'canceled']
canceled

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
0         505000  363965       mobile  canceled                           NaN   
3         508865  299063      connect  canceled                           NaN   
8         512475  322502       mobile  canceled                           NaN   
10        513743  330658       mobile  canceled                           NaN   
11        514161  366037      connect  canceled                           NaN   
...          ...     ...          ...       ...                           ...   
21283     569325  345079       mobile  canceled                           NaN   
21287     569764  405347       mobile  canceled                           NaN   
21288     570001  386413      connect  canceled                           NaN   
21297     571481  311841       mobile  canceled                           NaN   
21300     572293  355023      connect  canceled                           NaN   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
0                           NaN                                         NaN  
3                           NaN                                         NaN  
8                           NaN                                         NaN  
10                          NaN                                         NaN  
11                          NaN                                         NaN  
...                         ...                                         ...  
21283                       NaN                                         NaN  
21287                       NaN                                         NaN  
21288                       NaN                                         NaN  
21297                       NaN                                         NaN  
21300                       NaN                                         NaN  

[3265 rows x 7 columns]

In [73]:
positive_count = (data['delay_at_checkout_in_minutes'] > 0).sum()


non_nan_count = data['delay_at_checkout_in_minutes'].notna().sum()


negative_percentage = (positive_count / non_nan_count) * 100

print(f"Number of late checkout: {positive_count}")
print(f"Percentage of late checkout: {negative_percentage:.2f}%")

Number of late checkout: 9404
Percentage of late checkout: 57.53%


In [74]:
positive_values = data[data['delay_at_checkout_in_minutes'] > 0]

fig = px.histogram(
    positive_values['delay_at_checkout_in_minutes']/60,
    x='delay_at_checkout_in_minutes',
    nbins=1000,
    title="Distribution of Positive Values in 'delay_at_checkout_in_minutes'",
    labels={'delay_at_checkout_in_minutes': 'Delay at Checkout (Hours)'}
)

fig.show()

In [75]:
fig = px.histogram(
    positive_values['delay_at_checkout_in_minutes']/60,
    x='delay_at_checkout_in_minutes',
    nbins=10000,
    title="Distribution of Positive Values (0 to 6 Hours) in 'delay_at_checkout_in_minutes'",
    labels={'delay_at_checkout_in_minutes': 'Delay at Checkout (Hours)'},
)

fig.update_xaxes(range=[0, 6])
fig.show()

In [76]:
positive_non_nan = data[(data['delay_at_checkout_in_minutes'] < 0) & (data['state'].notna())]

canceled_count = (positive_non_nan['state'] == 'canceled').sum()
total_positive_non_nan_count = positive_non_nan.shape[0]
canceled_percentage = (canceled_count / total_positive_non_nan_count) * 100

canceled_count, canceled_percentage

(np.int64(1), np.float64(0.01466275659824047))

In [77]:
positive_non_nan

rental_id  car_id checkin_type  state  delay_at_checkout_in_minutes  \
1         507750  269550       mobile  ended                         -81.0   
5         511626  398802       mobile  ended                        -203.0   
6         511639  370585      connect  ended                         -15.0   
7         512303  371242       mobile  ended                         -44.0   
15        518282  364160      connect  ended                         -81.0   
...          ...     ...          ...    ...                           ...   
21279     568725  403265      connect  ended                        -159.0   
21285     569415  408598       mobile  ended                         -74.0   
21304     573322  376491      connect  ended                         -66.0   
21306     573790  341965       mobile  ended                        -337.0   
21308     574852  362531      connect  ended                         -76.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
1                           NaN                                         NaN  
5                           NaN                                         NaN  
6                      563782.0                                       570.0  
7                           NaN                                         NaN  
15                          NaN                                         NaN  
...                         ...                                         ...  
21279                       NaN                                         NaN  
21285                       NaN                                         NaN  
21304                       NaN                                         NaN  
21306                       NaN                                         NaN  
21308                       NaN                                         NaN  

[6820 rows x 7 columns]

In [78]:
previous = data[data['previous_ended_rental_id'].notna()]
previous_canceled = previous[previous['state'] == 'canceled']
previous_canceled.drop(columns=['delay_at_checkout_in_minutes'], inplace=True)

/tmp/ipykernel_201644/1551251009.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [79]:
previous_canceled 

rental_id  car_id checkin_type     state  previous_ended_rental_id  \
204       543768  374169      connect  canceled                  543010.0   
242       546160  352528      connect  canceled                  546578.0   
504       564627  341431       mobile  canceled                  552005.0   
637       568657  317378      connect  canceled                  566412.0   
669       516550  377700       mobile  canceled                  545076.0   
...          ...     ...          ...       ...                       ...   
21172     566228  390871      connect  canceled                  568465.0   
21197     567305  405564       mobile  canceled                  552222.0   
21230     569706  245154      connect  canceled                  558088.0   
21269     568049  381499      connect  canceled                  562174.0   
21272     568241  396409       mobile  canceled                  566136.0   

       time_delta_with_previous_rental_in_minutes  
204                                         210.0  
242                                         630.0  
504                                         150.0  
637                                         210.0  
669                                         720.0  
...                                           ...  
21172                                        60.0  
21197                                         0.0  
21230                                       660.0  
21269                                       720.0  
21272                                       570.0  

[229 rows x 6 columns]

In [80]:
previous_canceled = previous_canceled.merge(
    data[['rental_id', 'delay_at_checkout_in_minutes']],
    left_on='previous_ended_rental_id',
    right_on='rental_id',
    how='left'
)

previous_canceled.rename(columns={'delay_at_checkout_in_minutes': 'previous_delay_at_checkout'}, inplace=True)


In [81]:
previous_canceled

rental_id_x  car_id checkin_type     state  previous_ended_rental_id  \
0         543768  374169      connect  canceled                  543010.0   
1         546160  352528      connect  canceled                  546578.0   
2         564627  341431       mobile  canceled                  552005.0   
3         568657  317378      connect  canceled                  566412.0   
4         516550  377700       mobile  canceled                  545076.0   
..           ...     ...          ...       ...                       ...   
224       566228  390871      connect  canceled                  568465.0   
225       567305  405564       mobile  canceled                  552222.0   
226       569706  245154      connect  canceled                  558088.0   
227       568049  381499      connect  canceled                  562174.0   
228       568241  396409       mobile  canceled                  566136.0   

     time_delta_with_previous_rental_in_minutes  rental_id_y  \
0                                         210.0       543010   
1                                         630.0       546578   
2                                         150.0       552005   
3                                         210.0       566412   
4                                         720.0       545076   
..                                          ...          ...   
224                                        60.0       568465   
225                                         0.0       552222   
226                                       660.0       558088   
227                                       720.0       562174   
228                                       570.0       566136   

     previous_delay_at_checkout  
0                          42.0  
1                        -536.0  
2                         -88.0  
3                        -129.0  
4                         -35.0  
..                          ...  
224                         NaN  
225                         NaN  
226                       -40.0  
227                        73.0  
228                        16.0  

[229 rows x 8 columns]

In [82]:
delay_canceled = previous_canceled[previous_canceled['previous_delay_at_checkout'].notna()]


fig = px.histogram(
    delay_canceled['previous_delay_at_checkout']/60,
    x='previous_delay_at_checkout',
    nbins=1000,
    title="Distribution of Positive Values in 'delay_at_checkout_in_minutes'",
    labels={'delay_at_checkout_in_minutes': 'Delay at Checkout (Hours)'}
)

fig.show()

In [83]:
impact = previous_canceled[previous_canceled['previous_delay_at_checkout']>previous_canceled['time_delta_with_previous_rental_in_minutes']]

In [84]:
impact

rental_id_x  car_id checkin_type     state  previous_ended_rental_id  \
11        539151  282893       mobile  canceled                  548646.0   
16        553139  297851       mobile  canceled                  547650.0   
18        559037  349171      connect  canceled                  552392.0   
22        543745  359045      connect  canceled                  521852.0   
25        538224  340014      connect  canceled                  534999.0   
27        551372  261576       mobile  canceled                  547240.0   
33        568105  351546      connect  canceled                  564144.0   
38        546930  316449       mobile  canceled                  551866.0   
56        556126  396250       mobile  canceled                  548864.0   
64        564075  401387       mobile  canceled                  552061.0   
65        542207  368432       mobile  canceled                  556043.0   
66        542224  327336       mobile  canceled                  538452.0   
67        534020  384040      connect  canceled                  550585.0   
68        527015  243721       mobile  canceled                  534917.0   
74        549023  365926      connect  canceled                  550970.0   
85        556882  380479       mobile  canceled                  558724.0   
94        546592  296967      connect  canceled                  545025.0   
106       569087  193097      connect  canceled                  566731.0   
111       548079  364890       mobile  canceled                  547164.0   
116       550907  334190      connect  canceled                  549759.0   
117       545586  384396      connect  canceled                  553994.0   
122       530845  373889       mobile  canceled                  536992.0   
136       560412  336349      connect  canceled                  568759.0   
144       558830  353425      connect  canceled                  557068.0   
145       562555  380662       mobile  canceled                  567954.0   
150       572707  241021      connect  canceled                  574151.0   
158       544699  384193      connect  canceled                  529134.0   
170       550335  364199       mobile  canceled                  515060.0   
171       556230  387976       mobile  canceled                  553955.0   
172       539705  406161       mobile  canceled                  559887.0   
179       551166  342699      connect  canceled                  548950.0   
190       561559  390927      connect  canceled                  537960.0   
197       550972  360981       mobile  canceled                  548903.0   
198       552462  381787      connect  canceled                  548193.0   
207       568428  351660      connect  canceled                  563040.0   
211       571552  403962      connect  canceled                  574540.0   
223       566008  378689       mobile  canceled                  564756.0   

     time_delta_with_previous_rental_in_minutes  rental_id_y  \
11                                         30.0       548646   
16                                        240.0       547650   
18                                          0.0       552392   
22                                        150.0       521852   
25                                        180.0       534999   
27                                        210.0       547240   
33                                        120.0       564144   
38                                          0.0       551866   
56                                        510.0       548864   
64                                         60.0       552061   
65                                          0.0       556043   
66                                          0.0       538452   
67                                          0.0       550585   
68                                        120.0       534917   
74                                          0.0       550970   
85                                          0.0       

In [85]:
import math

total_count = data['state'].notna().sum()  
canceled_count = (data['state'] == 'canceled').sum()
canceled_percentage = (canceled_count / total_count) * 100

z_score = 1.96  
p = canceled_count / total_count  
n = total_count  
margin_of_error = z_score * math.sqrt((p * (1 - p)) / n) * 100  

canceled_percentage, margin_of_error

(np.float64(15.321445330830596), 0.4836163030240438)

Canceled reservations account for 15.3% of all the reservation of the data set with a mrgin error of 0.5% (95% confidence interval)

In [86]:
total_count = previous['state'].notna().sum()  
canceled_count = (previous['state'] == 'canceled').sum()
canceled_percentage = (canceled_count / total_count) * 100

z_score = 1.96  
p = canceled_count / total_count  
n = total_count  
margin_of_error = z_score * math.sqrt((p * (1 - p)) / n) * 100  

canceled_percentage, margin_of_error

(np.float64(12.438891906572515), 1.5075636194585913)

In [87]:
delta = data[data['time_delta_with_previous_rental_in_minutes'].notna()]
total_count = delta['state'].notna().sum()  
canceled_count = (delta['state'] == 'canceled').sum()
canceled_percentage = (canceled_count / total_count) * 100

z_score = 1.96  
p = canceled_count / total_count  
n = total_count  
margin_of_error = z_score * math.sqrt((p * (1 - p)) / n) * 100  

canceled_percentage, margin_of_error

(np.float64(12.438891906572515), 1.5075636194585913)

In [88]:
canceled

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
0         505000  363965       mobile  canceled                           NaN   
3         508865  299063      connect  canceled                           NaN   
8         512475  322502       mobile  canceled                           NaN   
10        513743  330658       mobile  canceled                           NaN   
11        514161  366037      connect  canceled                           NaN   
...          ...     ...          ...       ...                           ...   
21283     569325  345079       mobile  canceled                           NaN   
21287     569764  405347       mobile  canceled                           NaN   
21288     570001  386413      connect  canceled                           NaN   
21297     571481  311841       mobile  canceled                           NaN   
21300     572293  355023      connect  canceled                           NaN   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
0                           NaN                                         NaN  
3                           NaN                                         NaN  
8                           NaN                                         NaN  
10                          NaN                                         NaN  
11                          NaN                                         NaN  
...                         ...                                         ...  
21283                       NaN                                         NaN  
21287                       NaN                                         NaN  
21288                       NaN                                         NaN  
21297                       NaN                                         NaN  
21300                       NaN                                         NaN  

[3265 rows x 7 columns]

In [89]:
delta

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
6         511639  370585      connect     ended                         -15.0   
19        519491  312389       mobile     ended                          58.0   
23        521156  392479       mobile     ended                           NaN   
34        525044  349751       mobile     ended                           NaN   
40        528808  181625      connect     ended                         -76.0   
...          ...     ...          ...       ...                           ...   
21269     568049  381499      connect  canceled                           NaN   
21272     568241  396409       mobile  canceled                           NaN   
21275     568523  297973       mobile     ended                          12.0   
21286     569717  377312       mobile     ended                         230.0   
21305     573446  380069       mobile     ended                           NaN   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
6                      563782.0                                       570.0  
19                     545639.0                                       420.0  
23                     537298.0                                         0.0  
34                     510607.0                                        60.0  
40                     557404.0                                       330.0  
...                         ...                                         ...  
21269                  562174.0                                       720.0  
21272                  566136.0                                       570.0  
21275                  567121.0                                       240.0  
21286                  545045.0                                        90.0  
21305                  573429.0                                       300.0  

[1841 rows x 7 columns]

Becasue the percentag of canceled reservations in the subsets delay and previous is really close to value for the whole dataset both those subsets do not seems biased on the regard. It seems safe to draw conclusion from those subset of data for the whole dataset.

In [90]:
total_count = previous['state'].notna().sum()  
delay_canceled_count = (impact['state'] == 'canceled').sum()
delay_canceled_percentage = (delay_canceled_count / total_count) * 100

z_score = 1.96  
p = delay_canceled_count / total_count  
n = total_count  
margin_of_error = z_score * math.sqrt((p * (1 - p)) / n) * 100  

delay_canceled_percentage, margin_of_error

(np.float64(2.009777294948398), 0.6410538565669418)

In [ ]:
data = pd.read_excel('get_around_delay_analysis.xlsx')

In [21]:
data['delay_at_checkout_in_minutes'] = data['delay_at_checkout_in_minutes'].fillna(0)


In [22]:
data

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
0         505000  363965       mobile  canceled                           0.0   
1         507750  269550       mobile     ended                         -81.0   
2         508131  359049      connect     ended                          70.0   
3         508865  299063      connect  canceled                           0.0   
4         511440  313932       mobile     ended                           0.0   
...          ...     ...          ...       ...                           ...   
21305     573446  380069       mobile     ended                           0.0   
21306     573790  341965       mobile     ended                        -337.0   
21307     573791  364890       mobile     ended                         144.0   
21308     574852  362531      connect     ended                         -76.0   
21309     575056  351549      connect     ended                          35.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
0                           NaN                                         NaN  
1                           NaN                                         NaN  
2                           NaN                                         NaN  
3                           NaN                                         NaN  
4                           NaN                                         NaN  
...                         ...                                         ...  
21305                  573429.0                                       300.0  
21306                       NaN                                         NaN  
21307                       NaN                                         NaN  
21308                       NaN                                         NaN  
21309                       NaN                                         NaN  

[21310 rows x 7 columns]

In [68]:
threshold = 1

filtered_data = previous[(previous['time_delta_with_previous_rental_in_minutes'] < threshold)]
affected_rentals = len(filtered_data)
affected_rentals_pct = ((affected_rentals) /len(previous)) * 100

print(f"{affected_rentals} rentals are affected, which is {affected_rentals_pct:.2f}% of the total rentals.")

279 rentals are affected, which is 15.15% of the total rentals.


In [69]:
filtered_data

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
23        521156  392479       mobile     ended                           0.0   
107       537576  397470       mobile     ended                          18.0   
148       540479  374684       mobile     ended                          12.0   
164       541862  382364       mobile     ended                         125.0   
342       552330  384657      connect     ended                          23.0   
...          ...     ...          ...       ...                           ...   
21026     560998  412455       mobile  canceled                           0.0   
21054     562649  379751      connect     ended                          72.0   
21163     565721  381470       mobile     ended                          44.0   
21169     566008  378689       mobile  canceled                           0.0   
21197     567305  405564       mobile  canceled                           0.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
23                     537298.0                                         0.0  
107                    539005.0                                         0.0  
148                    539751.0                                         0.0  
164                    540607.0                                         0.0  
342                    554800.0                                         0.0  
...                         ...                                         ...  
21026                  561577.0                                         0.0  
21054                  565386.0                                         0.0  
21163                  564855.0                                         0.0  
21169                  564756.0                                         0.0  
21197                  552222.0                                         0.0  

[279 rows x 7 columns]

In [ ]:
total_reservations = len(data)

# Slider to set threshold
threshold = 
# Filter the data for delays greater than the threshold
affected_data = data[(data['delay_at_checkout_in_minutes'] > threshold)&(data['delay_at_checkout_in_minutes'] <= 0)]
affected_count = len(affected_data)
affected_ratio = (affected_count / total_reservations) * 100

# Display results
print(f"**Total reservations**: {total_reservations}")
print(f"**Reservations affected with threshold {threshold} minutes**: {affected_count}")
print(f"**Affected ratio**: {affected_ratio:.2f}%")

**Total reservations**: 21310
**Reservations affected with threshold 0 minutes**: 0
**Affected ratio**: 0.00%
